posts/2021/09/re-engineering-cicd-pipelines-for-scipy.md

```diff
@@ -0,0 +1,211 @@
+
+With the help of this, we were able to remove the hard-coded `__config__.py` and `version.py` from the source directory and ensure a clean build and install. [PR #57](https://github.com/rgommers/scipy/pull/57) contains the commits for the code required to generate the dynamic files and associated refactors in the `setup.py` file.
+
+## Integrating build caching using Ccache
+
+Today, it takes around ~35 minutes for SciPy's Linux and macOS tests to pass on GitHub Actions. With Scipy’s migration efforts towards the Meson build system, it takes around ~25 minutes for the CI to pass the build, install, and run tests. Our primary focus was to reduce the overall time to achieve this. But finding workarounds for the same was particularly challenging.
+
+John initially proposed the idea of caching the source and the build directory together. We can restore the cache during the workflow and use remote sync to update the subset of changed files. It would allow us to perform an incremental build.
+
+While discussing with Ralf, I realized that using timestamp-based caching would be the way to go forward. My initial approach was to cache the build directory, which was nowhere effective. My conversation with Meson developers helped me realize that using a compiler caching tool like `ccache` or `sccache` would be the way forward to cache the build targets.
+
+To make sure that dependencies are being cached, I capitalized on the [cache action](https://github.com/actions/cache). GitHub Workflow runs often reuse the same downloaded dependencies from one run to another. GitHub’s cache actions allow us to retrieve cache identified by a unique key, for which I used a unique timestamp-based caching approach.
+
+While digging around the documentation, I discovered that `ccache` is natively supported by Meson and hence would be a better choice. I found that the build targets are exactly being cached by debugging around, and I can use the cache action to cache them. Finally, I made it work, thus effectively reducing the build time by nearly ~70% and reducing the CI time.
+
+In addition, I also worked on caching PyPi dependencies using the cache actions that allowed us to shave off some of the time required during the Python dependency installation. [PR #63](https://github.com/rgommers/scipy/pull/63) containing the commits for the code necessary to set up build caching as part of the GitHub Action workflow for Linux.
+
+<p align="center">
+    <img
+     alt="Meson build system natively supports Ccache for compiler target caching. One of the project milestones required us to reduce the overall CI build time. With the help of compiler caching and the Cache action provided by GitHub Actions we were able to significantly tackle the total CI build time."
```

Comment:
Here, I think we need to emphasize the information you want to convey with the images: the difference in build time.